### PR TITLE
[Flux] Add advanced training script + support textual inversion inference

### PR DESCRIPTION
**Pull request by linoytsaban**
This PR adds an advanced version of the dreambooth lora flux script, with additional accompanying update to FluxPipline: 

-  adds an advanced script with pivotal tuning feature for CLIP & T5 encoder
- - `--train_text_encoder_ti` (enables CLIP pivotal tuning)
- - `--enable_t5_ti` adds T5 to the mix
- - `--train_text_encoder_ti_frac` the portion of epochs to train embeddings on, when using just clip
- - `--train_transformer_frac` the portion of epochs to train the transformer on, train_transformer_frac==0 will trigger a "pure_textual_inversion" train run (= "classical" textual inversion, no optimization of transformer lora layers)
- - `--initializer_token` the token to init the textual inversion embeddings with instead of randomly (random by default).
-  `--lora_blocks` the blocks\layers to apply lora training on. 

-  modifies Flux Pipeline (and related pipelines) to allow for textual inversion inference (that is also required by Loras trained with pivotal tuning)

_motivation:_ to allow for fast iterations & experimental features. I think it'd be good to fork the canonical script and bring the above changes into the advanced folder in a similar manner to what we have for SDXL.

cc @apolinario

### CODE REVIEW

**Comment by linoytsaban** [comment]
@sayakpaul @apolinario wdyt about me closing #9160 and moving the changes here as I suggest above?

**Comment by sayakpaul** [comment]
I would prefer #9160 as it helps to review the changes to the canonical script in isolation. Would that work for you?

**Comment by linoytsaban** [comment]
Yeah that makes sense, how about I adjust #9160 to be only for latent caching support for the canonical script? that way the remain isolated

**Comment by sayakpaul** [comment]
That works for me @linoytsaban!

File: examples/advanced_diffusion_training/README_flux.md
Comment (apolinario) [inline examples/advanced_diffusion_training/README_flux.md:84]: Here, if this is not a single token but rather it would take as many tokens as the person initializes, we probably should change the name to `intializer_concept`

File: examples/advanced_diffusion_training/README_flux.md
Comment (apolinario) [inline examples/advanced_diffusion_training/README_flux.md:288]: very nice!

**Comment by linoytsaban** [comment]
@sayakpaul  Yes ofc will add some examples soon! 
It is important to note that some of these are _very_ experimental and are meant for the community to explore and experiment with, with the motivation for all being integrating techniques (or modified versions of them) that showed success in dreambooth lora fine tuning of previous base models

**Comment by apolinario** [comment]
IMO we could merge the current model and add exmaples/samples built both by us and the community on a later stage, as the idea is less about bragging this changes have a big positive effect, and more to invite the community to test out the many possibilities this script brings (T5 fine-tuning, pivotal tuning with only CLIP, pivotal tuning with CLIP+T5, pure textual inversion, few different ways of doing token initialization). I think it's best to descentralize this experiments with the community vs running all ourselves (we're running them anyways though, but probably not needed as a bottleneck)

**Comment by sayakpaul** [comment]
Having the results apriori is tremendously beneficial. We don't have to be exhaustive in our hparam space here just 3/4 should suffice IMO.

**Comment by apolinario** [comment]
The idea of the advanced folder was a bit more loose and experimental and freeform - similar to the `community` folder, so I don't think we need to block merging before experiments, and as mentioned before, we'd like to share the experiments with the community. This said, Linoy is running a very extensive benchmarking of a lot of those params, but I don't see why bottlenecking.

**Comment by sayakpaul** [comment]
Sure but the PR doesn't supplement a single result gained from the additions. My understanding is that we have at least one result that we feel confident about regarding the advancements we're introducing in this PR. Or is that not a reasonable expectation? 

In that case, I think it's totally fine not to wait for even 3/4 of the experimental results before merging (which I still think is reasonable enough as an expectation).

**Comment by apolinario** [comment]
This is a bit different than other stuff: we are adding to the script techniques that are well used and work on SDXL: pivotal tuning, pure textual inversion, t5 text encoder training, and some other minor things. Those are tools on a toolbox and we want to put this tools out and invite the community to try/test/expand; The goal is *not* to provide results that we know that are improved with this additions, but rather that there are tools that have a big combinatorial analysis of possibilities and we want to invite people to test

I honestly don't see an issue with sharing this wht the community and bringing results/what worked/what doesn't on a later stage

**Comment by linoytsaban** [comment]
@sayakpaul Since there's quite a bit going on in this script, it would be great to get your thoughts if you could give the code a look in the meantime 🤗 

As poli mentioned I'm running multiple experiments on different uses cases (style, object, face) with different combinations of the new features so it takes a bit of time (that way we'll have some sort of benchmark of the contributions) but I'll add to this thread soon!

It'd be awesome if we can parallel this with the review as this will also allow for faster iterations in case there will be changes that require us to do more training runs

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:232]: Let's remove this.

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:214]: We have had a couple of issues when running the validation in between Flux LoRA training. So, let's make sure to verify this.

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:252]: Let's also garbage collect `gc.collect()`.

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:775]: Let's use `logger.warning()` for warnings to keep things consistent?

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:1127]: I am wondering if we should handle the prompt encoding part, leveraging the `encode_prompt()` method of `FluxPipeline`. If we can do this, it will help us in getting rid of a lot of boilerplate code that we're reusing from the pipeline. 

It would look something like this:

```python
text_encoder = CLIPTextModel.from_pretrained(
    ckpt_id, subfolder="text_encoder", torch_dtype=torch.bfloat16
)
text_encoder_2 = T5EncoderModel.from_pretrained(
    ckpt_id, subfolder="text_encoder_2", torch_dtype=torch.bfloat16
)
tokenizer = CLIPTokenizer.from_pretrained(ckpt_id, subfolder="tokenizer")
tokenizer_2 = T5TokenizerFast.from_pretrained(ckpt_id, subfolder="tokenizer_2")

pipeline = FluxPipeline.from_pretrained(
    ckpt_id,
    text_encoder=text_encoder,
    text_encoder_2=text_encoder_2,
    tokenizer=tokenizer,
    tokenizer_2=tokenizer_2,
    transformer=None,
    vae=None,
).to("cuda")

with torch.no_grad():
    print("Encoding prompts.")
    prompt_embeds, pooled_prompt_embeds, text_ids = pipeline.encode_prompt(
        prompt=prompt, prompt_2=None, max_sequence_length=256
    )
```

Perfectly fine if we completely get rid of functions like `load_text_encoders()`. WDYT?

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:1330]: Let's remove the unnecessary comments?

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:1527]: We are using a mix of `accelerator.print()`, `logger.info()`, and `print()`. Let's perhaps just stick to one?

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:1711]: Can be removed later?

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:1901]: Just a suggestion. Even though we want to provide options to the community, I think with a model like Flux, the use of prior preservation just really gets diminished, especially when we're doing text encoder fine-tuning as well as textual-inversion.

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:1969]: We could also do `pin_mememory=True` for non-blocking transfers. This way, we have a page-locked memory. 

https://discuss.pytorch.org/t/when-to-set-pin-memory-to-true/19723

**Review by sayakpaul — COMMENTED** [review]
Thanks! Most of my comments are pretty minor in nature except for trying to reuse the `encode_prompt()` method on the `FluxPipeline`. LMK if it makes sense.

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py
Comment (linoytsaban) [reply examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:1127]: the thing is that for pivotal tuning / textual inversion we need to mess with the tokenizers and text encoders both to add the tokens and access the embeddings and to then optimize so I'm not sure I see how that could work together

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py
Comment (linoytsaban) [reply examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:1901]: I tend to agree! my main reason why I would keep it is for multi-concept training where you could utilize prior preservation to maintain the knowledge of a previously _trained_ concept. WDYT? also maybe @apolinario has thoughts on this

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py
Comment (sayakpaul) [reply examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:1127]: Okay then disregard.

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py
Comment (sayakpaul) [reply examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:1901]: Okay if you feel so, we can keep it.

**Comment by sayakpaul** [comment]
@linoytsaban let me know when you need another review.

**Review by sayakpaul — APPROVED** [review]
Thanks!

**Comment by sayakpaul** [comment]
@yiyixuxu could you also give the changes in the pipelines a look? It's just about adding `TextualInversionMixin` so that we can enable pivotal tuning on Flux.

**Review by yiyixuxu — APPROVED** [review]
thanks!

**Review by apolinario — APPROVED** [review]
Let's go! 🚀

**Pull request merged**
